src/util/hass-call-api.js

```diff
@@ -1,28 +1,30 @@
-export default function hassCallApi(host, auth, method, path, parameters) {
-  var url = host + '/api/' + path;
+export default function hassCallApi(host, auth, method, path, parameters, responseType = '') {
+  const url = `${host}/api/${path}`;
 
   return new Promise(function (resolve, reject) {
-    var req = new XMLHttpRequest();
+    const req = new XMLHttpRequest();
+    req.responseType = responseType;
     req.open(method, url, true);
     req.setRequestHeader('authorization', `Bearer ${auth.accessToken}`);
 
     req.onload = function () {
-      let body = req.responseText;
-      const contentType = req.getResponseHeader('content-type');
+      let body = req.response;
+      if (!responseType) {
```

Comment:
This seems pretty weird. I think that if the response has a content type, we should be save to convert it to JSON. Otherwise, we can just return the body. It makes no sense to allow overriding the response type ?

Reply:
I'm not sure.  If the code requests a blob, it is expecting a blob to be returned.  If we actually got text and convert to json, the requesting code won't be happy.  To me the way it is makes sense.  Don't ask for a responseType unless you are prepared to handle it.  On the other hand, we shouldn't be requesting a blob if we are getting json back, and both the server and client are owned by homeassistant, so we are in control.

Reply:
Ah I see now the difference between `req.response` and `req.responseText`.

I don't think that I'm happy with this change, it seems like it becomes pretty unmaintainable. Note that building your own method is also unmaintainable.

The best solution is to create a `fetchWithAuth` method that calls APIs using `fetch()` and using the auth object to refresh the token if needed. This method is what you can use, as you have full access to the response object. `callApi` should then be rewritten to use `fetchWithAuth` under the hood.

That way there is 1 place that makes requests and people can use either as they see fit.

Reply:
I'll take a crack at it, but is there a reason to use 'fetch' vs 'XMLHttpRequest'?  Using fetch, I need to merger the passed 'init' variable to add new headers.  That means that either we pass back a modified init, or we need to copy it in the fetchWithAuth.  The copy must be a deep copy, and the 'headers' field is not gauranteed to be a dictionary.  In other words, there is quite a bit of processing needed to add the authorization header robustly in a fetchWithAuth function.  Building it around XMLHttpRequest is trivial in comparison

Reply:
```js
export const fetchWithAuth = async (auth, input, init = {}) => {
  if (auth.expired) await auth.refreshAccessToken();
  if (!init.headers) {
    init.headers = {};
  }
  init.headers.authorization = auth.accessToken;
  return await fetch(input, init);
};

```

Reply:
I think it's fine to mutate init, they are always created just for the fetch and not reused.